### ISSUE TITLE
Bug: [no-base-to-string] Maps result in [object Map] not [object Object]

### ISSUE THREAD

**Issue by kirkwaiblinger** [typescript-eslint/typescript-eslint#8606]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [X] I have tried restarting my IDE and the issue persists.
- [X] I have updated to the latest version of the packages.
- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22package%3A+eslint-plugin%22) and found none that matched my issue.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Playground Link

https://typescript-eslint.io/play/#ts=5.3.3&fileType=.ts&code=BQOwpg7gBAsghgB2ASmQOgC4HsDKGBOAliAOYoDcQA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1oCMBDZRLXxdk%2BKkwDm6KImjQO0SODABfECqA&tsconfig=&tokens=false

### Repro Code

```TypeScript
(new Map()).toString();
```

### ESLint Config

```javascript
module.exports = {
  "rules": {
    "@typescript-eslint/no-base-to-string": "error"
  }
}
```

### tsconfig

```jsonc
{
  "compilerOptions": {
    // ...
  }
}
```

### Expected Result

I expect it to flag with error message `'new Map()' will evaluate to '[object Map]' when stringified. 1:2 - 1:11`

### Actual Result

It flags with error message `'new Map()' will evaluate to '[object Object]' when stringified. 1:2 - 1:11`

### Additional Info

`new Set()).toString()` has the same nitpick.

**Comment by bradzacher** [comment]
The text is slightly wrong but the intent is correct - the `toString` is a bad one.

This behaviour in particular is driven by the existence of the `Symbol.toStringTag` property on a class. If it exists then the string of that property replaces `Object`.

However this property isn't declared on any of the types so it's impossible for us to know whether or not the "base toString" would return `[object Object]` or `[object Foo]`.

**Comment by JoshuaKGoldberg** [comment]
> this property isn't declared on any of the types

@bradzacher but we do know that it's an instance of the built-in object, no? We could do a check for the type of the instance to see if it is a `Map` or `Set`, I'd think.

**Comment by bradzacher** [comment]
@JoshuaKGoldberg I mean - sure - we could add custom handling here.
But doing so would be a decent chunk of code and effort.
Seems very low value just so we can tweak the error message in a way that (IMO) doesn't improve anything or clarify anything.

EG we can't just do "if type name is `Map` then `[object Map]`" because names aren't globally unique.
So instead we need to assert that the type name is `Map` and the type is defined in the TS lib.

But also what do you do in cases like `class Foo extends Map {}`? Well you need to inspect the prototype chain to see if there is a type named `Map` AND that type is from the TS lib types.
But also `Foo` might defined `Symbol.toStringTag` - do you handle that and not report instead?

I dunno - it seems like (a) the med-high effort required to implement (b) the low value gained in changing the message -- I'm a strong 👎 to this.

**Comment by JoshuaKGoldberg** [comment]
Yeah I'm not convinced it's worth the effort either. But I would be in support of changing the error message. Shall we retarget this issue to be for that?

**Comment by bradzacher** [comment]
> But I would be in support of changing the error message

What did you have in mind?

**Comment by JoshuaKGoldberg** [comment]
I ... can't think of anything great right now. @kirkwaiblinger do you have any ideas?

Proposal: let's leave this as `evaluating community engagement` to indicate we're waiting on a better error message? And if none comes along, we can close it out.

**Comment by bradzacher** [comment]
*Personally* I had no idea the `Symbol.toStringTag` behaviour existed or that it was implemented on the base types.

I'd hazard a guess that 99.999% of our users would be in that boat and wouldn't have ever (a) read the spec or (b) toString'd a Map.
Running a quick poll at work to confirm...

What I'm saying is that as a super power user you're aware of this and even noticed it. But the vast majority of users likely wouldn't even know this.. And for those that do know -- I doubt most would even put together "I'm stringifying a map so it'll say `Map` in the string" and "the error message says `Object`".

That's a long way of me explaining that I don't think changing the message to make it more wordy and "less certain" solely to work around this edge case is a good outcome, personally.

**Comment by kirkwaiblinger** [comment]
lol I am very appreciative to be ranked amongst the "super duper power users" ❤️ 

In any case, though, I would say that my position here is really just based on the fact that I get unreasonably worked up when people in general, and software frameworks/docs specifically, make imprecise or technically incorrect claims, even when the overall point is basically correct. So, my thought process in this case was 

1. See a function in my work codebase getting flagged by linter, literally this:

    ```typescript
    function makeParameterString() {
        const parameterMap = new Map();
        parameterMap.set('x', 'y');
        parameterMap.set('y', 'z');
        // set a whole bunch of parameters
        return parameterMap.toString();
    }
    ```
1. Think, wow what a brazen violation of the rule, they must know something I don't - is the rule correct here?
1. Stringify Map in the browser - does _not_ come out as "[object Object]" (though still useless). Believe me, I did not know in advance that it would be `[object Map]` instead of `[object Object]`
1. Think - hmm, well, the rule wasn't perfectly correct there, so, maybe there's situations where it legitimately does do something useful.
1. Spend 30 mins reading up on Object.prototype.toString and toStringTag.
1. Start looking through the node docs to figure out whether it magically stringifies differently in Node. It doesn't
1. Think, hmm, I doubted the rule for longer than I ought to have, just because it had this technically incorrect detail. 
1. Think, hey, I could probably fix that to save time for other people like me who are pathologically anal about details
1. Realize, oh, yeah, changing the code to actually report `[object Map]` is a way bigger pain than it's worth, but agree with discussion that it could be amelilorated by wording tweak
1. here we are, my case is made, call is up to maintainers to decide on

**Comment by kirkwaiblinger** [comment]
Oh, man, ok, sorry to keep this conversation going so long. I've just learned a new piece of information...
For stringifying Maps, apparently there is a standard method with `Object.fromEntries()` that I just found about (https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/fromEntries). 

So, in my mind, I would find it helpful to try to detect when a Map is being coerced to a string, and adjust the error message to say "Did you mean to use `Object.fromEntries({{name}})`? Presumably we either have functionality already to detect when an object is the built-in Map type, or will have to have it anyway soon, for some proposed rules around not using `Object.entries(map)`. So maybe we can make the marginal cost not huge.

**Comment by kirkwaiblinger** [comment]
Lmao I love the poll 🤣.

Yeah, see the thing is, once you learn that [object Map] is a thing, you immediately assume that [object MyType] will happen on a user defined class, and start to think, maybe that's not so bad, maybe these base-to-string warnings aren't all that useful. But No! JS is stupider than that!

**Comment by JoshuaKGoldberg** [comment]
I'm still not convinced one way or another... in large part because I haven't been in the mindset to truly and deeply think through this.

**Comment by Josh-Cena** [comment]
I'm +1 on an error message more nuanced than the current one, but I don't have great suggestions, maybe "It will use Object's default stringification format ([object Object]), which is likely not useful for display"? Vague but correct.

I'm -0.5 to test `@@toStringTag`` and change the error message based on that.

**Comment by JoshuaKGoldberg** [comment]
Aside: this might be the first issue promoted from `evaluating community engagement` to `accepting prs` we've ever had. Certainly the first I can remember. The process works! 🚀
